### PR TITLE
Midi

### PR DESCRIPTION
**Pull request by apmcleod**
midi.py is relatively straightforward.

test_midi.py is messier:
1. Where to put test.mid
2. Some code duplication in the last 2 methods.
3. There must be a better way to do line 75.
4. Some tests rely on pretty_midi being correct (ok I think?)
5. I haven't yet implemented (or tested) rounding. Wasn't sure where to put it (here or in some utils.py file)

### CODE REVIEW

File: mdtk/tests/test_midi.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import shutil
+
+import mdtk.midi as midi
+
+USER_HOME = os.path.expanduser('~')
+TEST_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_test_cache')
+MIDI_PATH = f"mdtk{os.path.sep}tests{os.path.sep}test.mid"
+
+def test_df_to_csv():
+    notes = []
```
Comment (JamesOwers) [inline mdtk/tests/test_midi.py]: You can:

```
notes = {
    'onset': [0, .3, .3],
    'track': [1, 1, 2],
    'pitch': [42, 40, 56],
    'dur': [1.5, 1.07, 1.11]
}
df = pd.DataFrame(notes)
```
if you'd like to use fewer lines. Your call - your way is perhaps clearer.

File: mdtk/tests/test_midi.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    
+    midi.df_to_csv(df, csv_name)
+    
+    assert os.path.exists(csv_name), ('No csv created.')
+    
+    # Check that notes were written correctly
+    with open(csv_name, 'r') as file:
+        for i, line in enumerate(file):
+            split = line.split(',')
+            assert float(split[0]) == notes[i]['onset'], ("Onset time of " +
```
Comment (JamesOwers) [inline mdtk/tests/test_midi.py]: If you were to make my change above to `notes`, you'd have to change these to `notes['onset'][i]` etc.

File: mdtk/tests/test_midi.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    for i, instrument in enumerate(m.instruments):
+        for note in instrument.notes:
+            midi_notes.append({'onset': note.start * 1000,
+                               'track': i,
+                               'pitch': note.pitch,
+                               'dur': (note.end - note.start) * 1000})
+    
+    df_notes = []
+    for index, df_note in df.iterrows():
+        # There must be a better way to do this
```
Comment (JamesOwers) [inline mdtk/tests/test_midi.py]: I think `df.to_dict('records')` does what you want: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_dict.html

File: mdtk/midi.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            dur: The duration of the note (offset - onset), in milliseconds.
+        Sorting will be first by onset, then track, then pitch, then duration.
+    """
+    midi = pretty_midi.PrettyMIDI(midi_path)
+    
+    notes = []
+    for index, instrument in enumerate(midi.instruments):
+        for note in instrument.notes:
+            notes.append({'onset': note.start * 1000,
+                          'track': index,
```
[the remark this answers is not in the document]
Comment (apmcleod) [reply mdtk/midi.py]: There's not really a way to ensure this as far as I know, since different datasets probably have different ways of treating each instrument.

There seems to be a [MIDI standard](https://en.wikipedia.org/wiki/General_MIDI#Program_change_events), but that refers to actual MIDI events, independent of track number (and this is not guaranteed to be consistent even within a single piece). There may also be, eg. 2 separate guitar parts but on different tracks. The closest thing to a guarantee I've seen is that track 10 tends to be percussion.

So in short: no, it doesn't ensure that, and it would likely be non-trivial (but possible) to do so.

File: mdtk/tests/test_midi.py (hunk 1 above)
Comment (apmcleod) [reply mdtk/tests/test_midi.py]: I'll stick with mine here for now. If it gets much longer, I can change this.

File: mdtk/tests/test_midi.py (hunk 2 above)
Comment (apmcleod) [reply mdtk/tests/test_midi.py]: Gotcha.

File: mdtk/tests/test_midi.py (hunk 3 above)
Comment (apmcleod) [reply mdtk/tests/test_midi.py]: Yup, that's the one!

File: mdtk/midi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (apmcleod) [reply mdtk/midi.py]: Ok, so it seems pretty_midi ([here](https://github.com/craffel/pretty-midi/blob/78e33ab1c24e1156529320ea127dd51d57f7ab99/pretty_midi/pretty_midi.py#L226)) separates by "instrument,track" pair.

There are some MIDI files which separate into tracks by voice (for example fugues). So those would come as separate "instruments" for us (which is probably what we want). I'd go for leaving this alone for now.

[the remark this answers is not in the document]
Comment (JamesOwers) [reply mdtk/midi.py]: Sure - I agree: we'll come back to this if required

File: mdtk/tests/test_midi.py (hunk 1 above)
Comment (JamesOwers) [reply mdtk/tests/test_midi.py]: No problem
